core/src/main/java/hudson/slaves/ComputerRetentionWork.java

```diff
@@ -47,31 +48,53 @@
     /**
      * Use weak hash map to avoid leaking {@link Computer}.
      */
-    private final Map<Computer, Long> nextCheck = new WeakHashMap<>();
+    private final Map<Computer, Long> checkAgainAfterTick = new WeakHashMap<>();
+    private final Map<Computer, Boolean> lastOnlineState = new WeakHashMap<>();
+
+    /**
+     * Administrators can trade elevated CPU usage from frequent agent checks for responsiveness to capacity needs.
+     *
+     * Per the docs of {@link RetentionStrategy#check}, it is OK to rechecked earlier or later than requested.
+     * Make the startup/teardown more responsive and use the requested minutes as multiples of the configured interval.
+     * Setting the interval to 60 seconds (which is the default), retains the original behavior and
+     *  one has to wait up-to one minute for an in-demand agent (with inDemandDelay=0min) to start.
+     * Setting the interval to 10 seconds provides a more responsive behavior towards load changes and
+     *  one has to wait up-to ten seconds for an in-demand agent (with inDemandDelay=0min) to start.
+     */
+    private static final long checkIntervalSeconds = SystemProperties.getLong(ComputerRetentionWork.class.getName() + ".checkIntervalSeconds", TimeUnit.MINUTES.toSeconds(1));
```

Comment:
This if poorly set can lead jenkins to perform poorly due to contention for the queue lock

Reply:
Agree. WDYT about putting a lower limit of three seconds in place?